rocky/crisis_room/models.py

```diff
@@ -41,10 +64,80 @@ class DashboardData(models.Model):
     )
 
     class Meta:
-        unique_together = [["dashboard", "position"], ["dashboard", "findings_dashboard"]]
+        constraints = [
+            models.UniqueConstraint(
+                name="unique dashboard position",
+                fields=["dashboard", "position"],
+                deferrable=models.Deferrable.DEFERRED,
+            ),
+            models.UniqueConstraint(
+                name="unique dashboard name", fields=["dashboard", "name"], deferrable=models.Deferrable.DEFERRED
+            ),
+        ]
 
     def __str__(self) -> str:
         try:
             return str(self.dashboard)
         except Dashboard.DoesNotExist:
             return super().__str__()
+
+    def clean(self) -> None:
+        if self.recipe and self.query:
+            raise ValidationError(_("You have to choose between a recipe or a query, but not both."))
+        if self.query and not self.query_from:
+            raise ValidationError(_("You have set a query and not where it is from. Also set the query_from."))
+        if not self.recipe and not self.query_from and not self.query:
+            raise ValidationError(_("DashboardData must contain at least a 'recipe' or a 'query_from' with a 'query'."))
+        return super().clean()
+
+    def update_position(self, move: str) -> None:
+        if move not in ("up", "down"):
+            return
+
+        old_position = self.position
+        new_position = self.position + (-1 if move == "up" else 1)
+
+        if 1 <= new_position <= 16:
+            try:
+                swap_item = DashboardData.objects.get(dashboard=self.dashboard, position=new_position)
+
+                # Swap positions and temporarily change position to 0 to avoid conflicts
+                with transaction.atomic():
+                    self.position = 0
+                    self.save(update_fields=["position"])
+                    swap_item.position = old_position
+                    swap_item.save(update_fields=["position"])
+                    self.position = new_position
+                    self.save(update_fields=["position"])
```

Comment:
I would make modifications to the fields first and then finally one single call to `self.save`